### PR TITLE
feat(mito2): introduce TWCS active window compaction

### PR DESCRIPTION
**Pull request by v0y4g3r**
I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## Refer to a related PR or issue link (optional)

Depends on #8989.

Related to #8995.

## What's changed and what's your intention?

This PR reduces TWCS rewrite amplification during ingestion while bounding read amplification and prioritizing the hottest data.

- Add independent file-count triggers for active and inactive TWCS windows while preserving the legacy trigger option.
- Resolve legacy TWCS trigger aliases consistently for table- and database-level options, including canonical ALTER behavior.
- Bound inactive-window convergence by the rewrite budget and determine the active window from the highest SST sequence.
- Add an active-window L1 safety trigger so accumulated L1 files can compact without enabling mixed compaction too early.
- Execute newer windows before older windows and cap regular picker batches by actual compaction parallelism. Serial automatic compaction therefore replans from the latest SST state after each output, so newly flushed active-window L0 files do not wait behind stale queued work.
- Preserve persisted legacy TWCS options across upgrades and verify them with a compatibility case.
- Propagate tournament-tree invariant failures through FlatMerge errors instead of panicking.

## Known limitations

- If multiple windows contain files with the same maximum sequence, the current `max_by_key()` selection marks only one of them active. The other equally recent windows can be treated as inactive and enter the more aggressive inactive compaction policy early.
- After ingestion stops, the final maximum-sequence window never transitions to inactive because the design has no wall-clock last-write tracking, inactivity timeout, or timer-driven compaction wake-up. The active L1 safety trigger and manual `COMPACT TABLE` are the available convergence paths.
- Inactive-window convergence is best effort rather than guaranteed. If no candidate satisfies the rewrite budget and progress constraints, the window remains fragmented instead of performing an unbounded mixed rewrite.
- Compaction already in flight is not preempted. With the default serial automatic compaction, newly flushed L0 files may wait for the current output, but not for additional outputs selected from the previous snapshot.

Validation:

- `cargo nextest run -p mito2 --retries 0` (1218 passed)
- `cargo sqlness bare -t alter_database`
- `cargo sqlness bare -t alter_table_options`
- `cargo run -p sqlness-runner -- compat --topology standalone --from-version v1.1.4 --test-filter twcs_active_window_options`
- `cargo clippy -p mito2 --all-targets --all-features -- -D warnings`
- `cargo fmt --all -- --check`

## PR Checklist

- [x] I have written the necessary rustdoc comments.
- [x] I have added the necessary unit tests and integration tests.
- [ ] This PR requires documentation updates.
- [x] API changes are backward compatible.
- [x] Schema or data changes are backward compatible.

### CODE REVIEW

File: src/mito2/src/compaction/twcs.rs [hunk 1]
```diff
@@ -358,6 +417,17 @@ impl Candidate {
...
 
+    /// Whether rewriting this candidate stays within the output file budget.
+    /// Inactive-window merges use this to bound write amplification when they
+    /// bypass the balance checks. An unset budget means no bound: the operator
+    /// removed the output size limit explicitly.
+    fn within_rewrite_budget(&self, max_output_file_size: Option<u64>) -> bool {
+        match max_output_file_size {
+            Some(limit) if limit > 0 => self.total_size <= limit as usize,
+            _ => true,
+        }
```
[the remark this answers is not in the document]
Comment (v0y4g3r) [reply src/mito2/src/compaction/twcs.rs:515]: GreptimeDB currently builds and releases only for x86_64, aarch64, and riscv64, where usize is 64 bits, so the narrowing described here cannot occur on a supported target. Candidate::absorb() also already converts file.size() from u64 to usize; changing only this comparison would not establish 32-bit correctness. No code change is needed for the supported targets.

File: src/sql/src/parsers/create_parser.rs [hunk 2]
```diff
@@ -223,9 +223,10 @@ impl<'a> ParserContext<'a> {
...
             .collect::<Result<HashMap<String, OptionValue>>>()?;
 
-        for key in options.keys() {
+        for (key, value) in &options {
             ensure!(
-                validate_database_option(key),
+                validate_database_option(key)
+                    && validate_database_option_value(key, value.as_string()),
                 InvalidDatabaseOptionSnafu { key: key.clone() }
             );
```
[the remark this answers is not in the document]
Comment (v0y4g3r) [reply src/sql/src/parsers/create_parser.rs]: Fixed in d29702a5ec. CREATE DATABASE now validates recognized option keys separately from their values and returns InvalidDatabaseOptionValue with the key, actual value, and constraint. Parser tests cover both L1 trigger keys, the valid boundary value 2, and the existing unknown-key diagnostic.

File: src/mito2/src/compaction/scheduler/planning.rs [hunk 3]
```diff
@@ -203,11 +203,14 @@ impl CompactionScheduler {
             )
         });
 
+        // Avoid queuing serial outputs from a stale snapshot. Repicking after each
+        // batch lets newly flushed L0 files outrank L1 work that has not started.
+        let max_picker_outputs = max_background_compactions.min(request.max_parallelism.max(1));
```
Comment (evenyag) [inline src/mito2/src/compaction/scheduler/planning.rs:208]: [P2] Preserve manual-compaction completion semantics when limiting batches

This cap also applies to manual regular compaction, whose default parallelism is 1. With two eligible windows, the request now processes only one before on_compaction_finished() notifies its waiter of success. Remaining work becomes an automatic follow-up that drops the requested time range; queued DDL can also suppress that follow-up. Previously, both windows could be selected and processed serially before completion.

Could we apply this cap only to automatic requests, or retain the manual waiter and requested time range across batches? A regression case with two eligible windows and manual parallelism 1 would cover this.

Comment (v0y4g3r) [reply src/mito2/src/compaction/scheduler/planning.rs:208]: This seems to affect only manual TWCS compaction. I’ll create an issue for it. The compaction scheduler needs a comprehensive refactor to distinguish the priorities of the following tasks:
- Manual compaction (and follow-ups)
- Automatic compaction for recent windows
- Automatic compaction for stale windows
Different tasks should also be assigned different rate limits and write cache admission.

File: src/common/meta/src/ddl/alter_database.rs [hunk 4]
```diff
@@ -51,7 +60,18 @@ fn build_new_schema_value(
...
+                            value.extra_options.remove(TWCS_TRIGGER_FILE_NUM);
+                            value
+                                .extra_options
+                                .remove(TWCS_ACTIVE_WINDOW_TRIGGER_FILE_NUM);
+                            TWCS_TRIGGER_FILE_NUM
+                        } else {
+                            key
+                        };
+                        value
+                            .extra_options
```
Comment (evenyag) [inline src/common/meta/src/ddl/alter_database.rs:75]: Can we document why we need this?

File: src/mito2/src/region/options.rs [hunk 5]
```diff
@@ -356,9 +369,22 @@ impl Default for CompactionOptions {
 #[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
 #[serde(default)]
 pub struct TwcsOptions {
-    /// Minimum file num in every time window to trigger a compaction.
+    /// Minimum file num in the active time window to trigger a compaction.
     #[serde_as(as = "DisplayFromStr")]
+    #[serde(rename = "active_window.trigger_file_num", alias = "trigger_file_num")]
     pub trigger_file_num: usize,
```
Comment (evenyag) [inline src/mito2/src/region/options.rs]: Do we need to rename the field to active_window_trigger_file_num?

Comment (v0y4g3r) [reply src/mito2/src/region/options.rs]: Renamed in 6d87268923. The Rust field is now `active_window_trigger_file_num`; the serde `rename` and legacy `alias` remain unchanged, so the persisted configuration format is preserved.

File: src/common/meta/src/ddl/alter_database.rs (hunk 4 above)
Comment (v0y4g3r) [reply src/common/meta/src/ddl/alter_database.rs:75]: Documented in 94b005f338. The comment now explains that we persist the legacy key so older versions can read the option after a downgrade, and that persisting both aliases would deserialize as a duplicate field.

**Review by evenyag — APPROVED** [review]
